examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts

```diff
@@ -86,100 +108,132 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
-                this.emit("migrationNotSupported", acceptedVersion);
-                this._migrationP = undefined;
-                return;
-            }
-            await migratedModel.importData(transformedData);
+                await migratedModel.importData(transformedData);
 
-            // Before attaching, let's check to make sure no one else has already done the migration
-            // To avoid creating unnecessary extra containers.
-            if (migratable.migrationTool.migrationState === "migrated") {
-                this._migrationP = undefined;
-                migratedModel.close();
-                this.takeAppropriateActionForCurrentMigratable();
-                return;
-            }
+                // Store the detached model for later use and retry scenarios
+                this._preparedDetachedModel = detachedModel;
+            };
 
-            // TODO: Support retry
-            // TODO: Use TaskManager here to reduce container noise.  Specifically -- all clients should race up to
-            // this point (so they're all prepared to do the migration) but should wait for the task lock before
-            // attempting the attach() to minimize the chance that multiple containers are created on the service.
-            const containerId = await createResponse.attach();
+            const completeTheMigration = async () => {
+                assert(this._preparedDetachedModel !== undefined, "this._preparedDetachedModel should be defined");
 
-            // Again, it could be the case that someone else finished the migration during our attach.
-            // Casting to MigrationState because TS doesn't understand that the state may have changed during the
-            // above await.
-            if (migratable.migrationTool.migrationState as MigrationState === "migrated") {
-                this._migrationP = undefined;
-                migratedModel.close();
-                this.takeAppropriateActionForCurrentMigratable();
-                return;
-            }
+                // Volunteer to complete the migration.
+                let isAssigned: boolean;
+                try {
+                    isAssigned = await this.currentModel.migrationTool.volunteerForMigration();
+                } catch (error) {
+                    // volunteerForMigration() will throw an error on disconnection. In this case, we should exit and
+                    // re-enter the state machine which will wait until we reconnect.
+                    // Note: while we wait to reconnect it is possible that another client will have already completed
+                    // the migration.
+                    assert(!this.connected, "We should be disconnected");
+                    return;
+                }
 
-            // TODO: Support retry
-            await migratable.migrationTool.finalizeMigration(containerId);
+                if (!isAssigned) {
```

Comment:
It's possible that another client successfully finalized the migration (wrote the new ID to the CRC) but disconnected/died before they were able to send the "complete" op -- if this happens we should avoid proceeding here.  I think this means we should check the migrationTool.newContainerId rather than trusting the isAssigned result here.

Reply:
Makes sense. We should also validate, probably with an assert, that `isAssigned` is true if we make it past that check